### ISSUE TITLE
What if it was only actions?

### ISSUE THREAD

**Issue by jbucaran** [hyperapp/hyperapp#126]
I am rewriting parts of the documentation and currently revisiting the eternal reducers Vs effects.

Why do we need to force users to think about reducers and effects instead of just **actions**? 

If an action returns a promise, we know it's an effect, otherwise it's a reducer.

```jsx
app({
  model: {
    counter: 0,
    waiting: false
  },
  actions: {
    add: model => ({ counter: model.counter + 1 }),
    toggle: model => ({ waiting: !model.waiting }),
    waitThenAdd: (model, actions) => {
      actions.toggle()
      return new Promise(resolve => setTimeout(_ => resolve(), 1000))
        .then(actions.add)
        .then(actions.toggle)
    }
  },
  view: (model, actions) =>
    <button
      onclick={actions.waitThenAdd}
      disabled={model.waiting}
    >
      {model.counter}
    </button>
})
```

The only drawback is that Promises are not supported in IE11, so obviously we can't change this right now.

**Comment by leeoniya** [comment]
Promise shims come pretty small: https://github.com/taylorhakes/promise-polyfill/blob/master/promise.min.js

**Comment by jbucaran** [comment]
@leeoniya Thanks! There must be an even smaller implementation that targets only IE10 or IE11, which is what HyperApp supports.

**Comment by jbucaran** [comment]
Alternatively, we could provide this out of the box like this:

```jsx
app({
  model: {
    counter: 0,
    waiting: false
  },
  actions: {
    add: model => ({ counter: model.counter + 1 }),
    toggle: model => ({ waiting: !model.waiting }),
    waitThenAdd: (model, actions) => {
      actions.toggle()
      return resolve => setTimeout(_ => resolve(actions.add).then(actions.toggle), 1000)
    }
  },
  view: (model, actions) =>
    <button
      onclick={actions.waitThenAdd}
      disabled={model.waiting}
    >
      {model.counter}
    </button>
})
```

**Comment by selfup** [comment]
The problem with setTimeout is that we aren't sure how long the request is going to take

**Comment by jbucaran** [comment]
@selfup See my edit. HyperApp receives the function you return from the effect and calls it immediately with our own home-brewed `then` function. 

setTimeout is called and eventually our then is called.

**Comment by selfup** [comment]
Ohh I see! Hmm using Events?

**Comment by selfup** [comment]
I was going to mention Proxy but no IE support. It seems the polyfill is small though:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

**Comment by jbucaran** [comment]
Here is a working POC:

```jsx
  Object.keys(app.actions).forEach(function (key) {
    actions[key] = function (data) {
      var result = app.actions[key](model, data, actions)
      if (typeof result === "function") {
        result(function then(f) {
          f()
          return { then: then }
        })
      } else {
        var oldModel = model
        model = merge(model, result)
        render(model, view)
      }
    }
  })
```

**Comment by selfup** [comment]
Oh I like that!

**Comment by jbucaran** [comment]
This would be quite a drastic change. Could I be missing a use case for having both reducers and effects as a separate unit?

Thoughts? @dodekeract @lukejacksonn @tzellman @danigb

**Comment by liadbiz** [comment]
long time after the last time I have been here. The project changed a lot! Thanks for the hard work of you @jbucaran  and other cool contributers.

But I think it "may" not be a good idea to mix both reducers and effects into actions.but it actually make hyperApp less concepts, and I like "less" concepts.

ah...it's hard to decide= =

**Comment by dodekeract** [comment]
@liadbiz I think it just hides the concepts, not reduces them. IMO this makes it less understandable and more limited.

For example I'm using some effects that do **synchronous** stuff like modifying some WebRTC SDP or just calling multiple other actions.

I think this makes hyperapp both more limited and harder to understand. :-1:

**Comment by jbucaran** [comment]
@liadbiz From a documentation point of view, we would continue using the term actions.

**Comment by jbucaran** [comment]
> @dodekeract For example I'm using some effects that do synchronous stuff like modifying some WebRTC SDP or just calling multiple other actions.

Sync effects, sure.

```jsx
actions: {
  mySyncEffect: (model, data, actions) {
    actions.multiple()
    actions.other()
    actions.actions()
  }
}
```

**Comment by jbucaran** [comment]
> I think this makes hyperapp both more limited and harder to understand.

If this is true, I'd like to see some proof.

1. How does it make hyperapp more limited? 
For example, provide an example of something that would be impossible using only actions, instead of the current architecture.

2. What's harder to understand? 
Imagine this is the first time you encountered hyperapp, how is learning only _actions_ more complicated than learning about "reducers" and "effects".

In fact, understanding the difference between reducers and effects is a common pitfall when learning Elm.

**Comment by jbucaran** [comment]
> @dodekeract I think this makes hyperapp both more limited and harder to understand.

If this is true, I'd like to see some proof.

**1. How does it make hyperapp more limited? **
For example, provide an example of something that would be impossible using only actions, instead of the current architecture.

**2. What's harder to understand? **
Imagine this is the first time you encountered hyperapp, how is learning only _actions_ more complicated than learning about "reducers" and "effects".

In fact, understanding the difference between reducers and effects is a common pitfall when learning Elm.

**Comment by jbucaran** [comment]
This would play well along async functions too.

**Comment by jbucaran** [comment]
> @liadbiz But I think it "may" not be a good idea to mix both reducers and effects into actions.but it actually make hyperApp less concepts, and I like "less" concepts.

For the record, we _are_ already mixing reducers and effects into actions. That's the `actions` object we pass to effects and subscriptions.
* https://github.com/hyperapp/hyperapp/wiki/api#effects

**Comment by jbucaran** [comment]
This also gives users the freedom to call reducers whatever they want, mutations, updaters, etc. /cc @SkaterDad

**Comment by traducer** [comment]
I really like having effects and reducers separate. At least to me the distinction between the two is very clear (async/sync). This may be because I'm more exposed to this pattern but it allows me to write code explicitly. If you combined them into just actions, I would see no real problem as they both will accomplish the same thing, but personally I like having them separate.

**Comment by jbucaran** [comment]
@traducer All I see is more words you have to learn or remember. In the end, the object you are using to trigger actions is called `actions`, so from that POV it makes sense to call the whole thing `actions` from the beginning.

Now, with nested actions, you have the freedom to create the.schema.of.your.dreams, so technically, what I am proposing is more alike moving these concepts out into userland. 

Out of the box, HyperApp would be like, I just know actions. Most people will just do fine and make some progress in their apps and at some point they will find they need to do something async so they will wonder: how can I make an action async?

Well, JavaScript has promises, maybe return one? Then they'll see in the docs that's exactly what they have to do and be like: I knew it!

And if you can't use promises because "you know why", then return a function with the signature `(then)`. This `then` function is provided by HyperApp and works as described above, basically like Promise.then.

**Comment by liadbiz** [comment]
@jbucaran 
this is actually what I am wondering.
what the "less" concepts that I have talking about means it's easy for newcommer to get started. 
At the same times, it "maybe" a good idea for experienced developer who are working on a complex project, nodoutedly, mixed reducer functions and effect functions into actions is not a good practice. 

however, it's all user choice, If user can choose the way they write the code, this idea could be appreciated. anyway, :+1:

**Comment by jbucaran** [comment]
> @liadbiz mixed reducer functions and effect functions into actions is not a good practice

Perhaps you are under the impression that people categorize actions based in which ones are sync or async?

I can imagine someone might do that, but in a real world situation, it's more likely you'll be organizing your tasks based in other aspects of your application, and not whether something is sync/async. 

If this feature lands, you will, however, be able to create a nested namespace inside `actions`, for example: `tasks`, `async`, `effects`. You'll be covered.

**Comment by liadbiz** [comment]
@jbucaran 
yeah, I guess that some people will like that way, such as users who have used React,  so I use "maybe". I personnaly do not have  a habbit  like this.

**Comment by liadbiz** [comment]
I have missed a lot during these busy days. Dont mind if understand something in a wrong way :) 
I will have a look at the wiki and source code tonight to grab more things.

**Comment by jbucaran** [comment]
@liadbiz No probs! Now that plugins have landed, I'll ship actions (you can see a WIP [here](https://github.com/hyperapp/hyperapp/pull/127)) and then release a new version.

This should be a major, but I am still going to delay 1.0 until after the entire docs and some better examples are available.


* https://github.com/hyperapp/hyperapp/pull/127

**Comment by nichoth** [comment]
Is returning a promise/function necessary? Could you just call multiple other actions and return undefined?

**Comment by dodekeract** [comment]
@nichoth it is necessary in the context of `async function`s.

E.g.

```javascript
effects: {
	example: async () => {
		await actions.login()
		await actions.getSession()
		await actions.loadProfile()
		actions.router.go('/profile')
	}
}
```

**Comment by nichoth** [comment]
@dodekeract hmm I don't see that it's preferable instead of using just callbacks. Is it a matter of style/preference?

**Comment by dodekeract** [comment]
> I see a ton of words you have to learn and remember
— @jbucaran

The way to get rid of those words is not to just hide them. The words are not the issue, the concepts are.

Since I believe that the concepts "effect" and "reducer" are the simplest things that can get the task done, so I don't think removing the words is a good thing to do.

**Comment by dodekeract** [comment]
@nichoth callbacks don't short-circuit the logic hyperapp uses. Using them would be a hack with the way the actions are written.

> I don't see that it's preferable

See [callback hell](http://callbackhell.com). IMO `async`/`await` is the best solution that came up so far to make async functions great again.

**Comment by nichoth** [comment]
Ok, it sounds like it is a style thing. I generally prefer callback hell over new language syntax.

**Comment by dodekeract** [comment]
@nichoth Try using hyperapp's actions with callbacks. If you manage to make it work it's actually impressive, especially when this proposal gets implemented.

**Comment by lukejacksonn** [comment]
This is a really interesting proposal and something that I have certainly found myself asking.

I have found in many cases an effect is just the promise of an update (do this async thing then call a reducer). In that vein, it could make sense to expect all actions to be a promise and resolve them to produce a model proposition that gets merged with model. Something like..

```
const actions = {
  fetchUser: (m,d,a) =>
    fetch(`https://api.something.com/item/${d}`)
    .then(data => data.json())
    .then(a.setUser)
  setUser: (m,d,a) => 
    ({ user: d })
}
```
Could you not do something like this..
```
Object.keys(app.actions).forEach(function (key) {
  actions[key] = function (data) {
    Promise.resolve(app.actions[key](model, data, actions))
    .then(result => {
      var oldModel = model
      model = merge(model, result || {})
      render(model, view)
    })
  }
})
```
Which allows the developer to do..
```
<button onclick={ actions.setUser({name: 'Joe Bloggs'}) }>SET</button>
<button onclick={actions.fetchUser(1)}>FETCH</button>
```

**Comment by nichoth** [comment]
@dodekeract Isn't that how it works currently? You can just call multiple actions from within an action, don't return anything.

```js
effects: {
    asyncThing: function (state, actions) {
        actions.startRequest()
        fetch(function (err, resp) {
            if (err) return actions.error()
            actions.resolve()
            actions.setState(resp)
        }
    }
}
```

**Comment by dodekeract** [comment]
@nichoth Yes, you can do that. Your actions themself won't be able to have all the necessary logic in them though.

Look at my example again. To get a user's profile you need to be signed in. So you **have to wait** until `login` is done.

Now you **could** duplicate the code and make a new action that does both and only use callbacks. Technically that would work. However you can easily compose multiple asynchronous actions by separating all api calls into one action per call.

That way you won't have to repeat the same `login` code again and again.

Of course it's possible to make all of this work by extracting parts of the codebase out of effects/reducers, but I don't think it's desirable or better in most situations.

So to recap - with your callback suggestion you would have trouble composing multiple async actions and you'd end up either duplicating or writing more unrelated code into the relevant action. (in my case `example`)

**Comment by dodekeract** [comment]
@lukejacksonn look at the example I just tried to explain to @nichoth.

`await action.blah` wouldn't be possible when hyperapp uses the promise itself.

**Comment by dodekeract** [comment]
Actually it's worth considering, but hyperapp would have to wrap it into another promise so `async`/`await` is still possible.

Biggest downside I see so far is that it'd make it harder to reason about your code and that you'd have to understand promises to use hyperapp properly.

**Comment by lukejacksonn** [comment]
You could just return the `Promise.resolve().then()` that would allow you to use the `await` syntax wouldn't it (see updated comment code)?

**Comment by lukejacksonn** [comment]
It would mean that all actions are made `then`-able even reducers (which gets turned into a promise that resolves immediately). This would be more consistent for the developer so something like..
```
actions.setUser({ name: 'Joe Bloggs' })
.then(actions.fetchUser(1))
```
is just as valid as..
```
actions.fetchUser(1)
.then(setUser({ name: 'Joe Bloggs' }))
```
Which could be written with `async`/`await` syntax. I'm probably missing something here but that is how I imagined this going!

**Comment by jbucaran** [comment]
Hey everyone! This issue is kinda out of sync and it also has some confusing comments I made while I was still thinking about how to implement this feature. Please see the updated comment and explanation given in #127 for the most up to date and correct information about this feature.

I'll try to address some of your comments:

> Is returning a promise/function necessary? Could you just call multiple other actions and return undefined?

**Nope**. You can continue using hyperapp as usual, except that instead of reducers/effects, the property name will be actions. What's new then? Now you can also return a promise however, and hyperapp will pass it to the caller letting you create a .then chain. 

> Ok, it sounds like it is a style thing.
 
Yes, this change is part a style thing and part an attempt to simplify the architecture.

> You can just call multiple actions from within an action, don't return anything.

Yep, you can do this right now, and you will be able to do so too after #127.

> In in many cases an effect is just the promise of an update

After #127 you will be continue to have actions that do async stuff in the same way you do it right now. You will also be able (if you like and this is optional) to return a Promise and hyperapp will pass it to the caller allowing `async/await` stuff or who knows what else.

/cc @nichoth @dodekeract @lukejacksonn

**Comment by jbucaran** [comment]
> @dodekeract Also it'd be really easy to mess something up and accidentally rewrite the whole model with garbage by sometimes resolving weirdly/returning the wrong promise/not handling errors properly/etc.

Yes, which is why it none of this is stuff is required.

**Comment by selfup** [comment]
The problem with that is that Promises are not very performant even if they resolve immediately

**Comment by jbucaran** [comment]
> @selfup The problem with that...

What is that?

**Comment by jbucaran** [comment]
> The problem with that is that Promises are not very performant even if they resolve immediately

I am not sure what was meant by *that*, but for the record, hyperapp will **not** require you to return a Promise, you could if you want and hyperapp will pass it to the caller (allowing async/await), but you **don't have** to. Basically do your async stuff inside the action (as usual) and when you're done call other actions (or not).

**Comment by selfup** [comment]
@jbucaran I think I was referring to @lukejacksonn's example of everything being a promise.

I guess by the time I hit submit you had already made some comments 😄

**Comment by lukejacksonn** [comment]
What @jbucaran is presuming about my proposal `that ever action would have to be a promise` is false.. I have answered both responses in the discussion to https://github.com/hyperapp/hyperapp/pull/127

**Comment by jbucaran** [comment]
@lukejacksonn Gotcha. But it is true that we'd had to use `Promise` inside hyperapp, which would make Hyperapp useless without a polyfill out of the box.

**Comment by jbucaran** [comment]
Going to close this as https://github.com/hyperapp/hyperapp/pull/127 was merged.

Please see the PR for the most up-to-date explanation of the feature.

Now, I need to update docs and examples. 🔥

**Comment by tunnckoCore** [comment]
d'oh
